bn_mp_factor.c

```diff
@@ -0,0 +1,217 @@
+#include "tommath_private.h"
+#ifdef BN_MP_FACTOR_C
+/* LibTomMath, multiple-precision integer library -- Tom St Denis
+ *
+ * LibTomMath is a library that provides multiple-precision
+ * integer arithmetic as well as number theoretic functionality.
+ *
+ * The library was designed directly after the MPI library by
+ * Michael Fromberger but has been written from scratch with
+ * additional optimizations in place.
+ *
+ * SPDX-License-Identifier: Unlicense
+ */
+
+#ifdef LTM_USE_EXTRA_FUNCTIONS
+/* Factor integer. Good for up to about 35-40 bit large factors */
+static int mp_factor_intern(mp_int *z, mp_factors *factors)
+{
+
+   int e = MP_OKAY;
+   int all, ilog2;
+   /* int perfpow; */
+
+   mp_int factor, rootout, exponent, quot, rem, r;
+
+
+   ilog2 = mp_count_bits(z);
+   /* sizeof(int) < 4 is possible (e.g.: for MP_8BIT/MP_16BIT) */
+   if (ilog2 < 10) {
+      ilog2 = 10;
+   } else if (ilog2 > (int)(2 * (sizeof(int) * CHAR_BIT) - 1u)) {
+      ilog2 = (int)((2 * sizeof(int) * CHAR_BIT) - 1u);
+   }
+
+   /* floor(sqrt(z)) */
+   ilog2 = (int)(1uL << (ilog2/2));
+
+#ifdef MP_8BIT
+   ilog2 = ((ilog2 > 0) && (ilog2 < (int)(LTM_SIEVE_UINT_MAX)))?ilog2:(int)(LTM_SIEVE_UINT_MAX);
+#else
+   ilog2 = ((ilog2 > 0) && (ilog2 < (int)(LTM_SIEVE_UINT_MAX_SQRT)))?ilog2:(int)(LTM_SIEVE_UINT_MAX_SQRT);
+#endif
+
+   if ((e = mp_init_multi(&factor, &rootout, &exponent, &quot, &rem, &r, NULL)) != MP_OKAY) {
+      return e;
+   }
+   /* Trial division */
+   if ((e = mp_trial(z, ilog2, factors, &r)) != MP_OKAY) {
+      goto LTM_ERR;
+   }
+
+   /* Check remainder to see if we are done here */
+   if (mp_cmp_d(&r,1uL) == MP_EQ) {
+      goto LTM_END;
+   } else {
+      /* Try to find more factors until the remainder is 1 (one)*/
+      do {
+         /*
+            Check if the remainder is already prime.
+            mp_prime_is_prime() uses BPSW which is tested up to 2^64 (~10^19)
+            but that number is already quite large for a factor.
+          */
+         if ((e = mp_prime_is_prime(&r, 8, &all)) != MP_OKAY) {
+            goto LTM_ERR;
+         }
+
+         if (all == MP_NO) {
+#if 0
```

Comment:
can this code go or is there a reason this should stay?

Reply:
The first one is useful for some numbers but will hold up everything for others and I'm benchmarking the other one if it is useful at that place. I'm thinking about adding a flag for the first one to let the caller decide. The latter is needed for the quadratic sieve, but that will come later, much later.

So: should it stay or should it go? Good question. (Why am I hearing The Clash in my head now?)

Oh, nearly forgot: both, `mp_isperfpower` and `mp_ispower` need a `mp_n_root_ex` with a normal runtime to begin with. It does not need the optimized version I have in a different branch (`nthroot` I think?) but at least the one I put into this PR. It implements just the standard seed (a bit above the actual root) and a slightly more paranoid check at the end. The usability of the original version is so small, I would consider the one in this PR nothing more than a bug-fix.